File: secure-gateway-studio/backend/src/sgstudio/providers/acceptance.py
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any
from urllib.parse import quote, urlsplit

from sgstudio.domain.canonical import canonical_json
from sgstudio.domain.models import (
    AcceptanceStatus,
    AcceptanceTestId,
    BackendKind,
    CertificateStrategy,
    DeploymentMode,
    DeploymentSpec,
)
from sgstudio.domain.planner import canonical_configuration_hash
from sgstudio.providers.google_rest import (
    GoogleApiError,
    GoogleAuthorizedTransport,
    JsonTransport,
)


@dataclass(frozen=True)
class AcceptanceFinding:
    test_id: AcceptanceTestId
    status: AcceptanceStatus
    summary: str
    evidence: str


class GoogleAcceptanceVerifier:
    """Verify deploy-time acceptance controls without collecting credentials."""

    def __init__(self, transport: JsonTransport) -> None:
        self._transport = transport

    def verify(self, spec: DeploymentSpec) -> list[AcceptanceFinding]:
        if spec.backend_kind is BackendKind.DIRECT_HTTPS:
            return [self._verify_application(spec)]
        if spec.backend_kind is BackendKind.INTERNAL_HTTPS_LB:
            return [
                self._verify_backend(spec),
                self._verify_dns(spec),
                self._verify_application(spec),
            ]
        findings: list[AcceptanceFinding] = []
        if spec.backend_kind is BackendKind.MANAGED_SAMPLE:
            findings.append(self._verify_backend(spec))
        findings.extend(self._verify_offload(spec))
        findings.append(self._verify_dns(spec))
        findings.append(self._verify_application(spec))
        return findings

    def _verify_backend(self, spec: DeploymentSpec) -> AcceptanceFinding:
        try:
            evidence = self._guest_attribute(
                spec,
                zone=spec.zone,
                instance=f"{spec.name}-backend",
                test_id=AcceptanceTestId.T01,
            )
            passed = (
                evidence.get("status") == 200
                and self._has_hash(evidence)
                and self._matches_configuration(evidence, spec)
            )
            return self._finding(
                AcceptanceTestId.T01,
                passed,
                "Managed backend returned HTTP 200 to its local runtime probe",
                "Managed backend runtime probe is missing or invalid",
                evidence,
            )
        except (GoogleApiError, ValueError, KeyError, TypeError):
            return self._failed(
                AcceptanceTestId.T01,
                "Managed backend runtime probe could not be verified",
                "guest-attribute-unavailable",
            )

    def _verify_offload(self, spec: DeploymentSpec) -> list[AcceptanceFinding]:
        try:
            instances = self._offload_instances(spec)
            if not instances:
                raise ValueError("no-offload-instances")
            t02_evidence = [
                {
                    "instance": instance,
                    **self._guest_attribute(
                        spec,
                        zone=zone,
                        instance=instance,
                        test_id=AcceptanceTestId.T02,
                    ),
                }
                for zone, instance in instances
            ]
            t03_evidence = [
                {
                    "instance": instance,
                    **self._guest_attribute(
                        spec,
                        zone=zone,
                        instance=instance,
                        test_id=AcceptanceTestId.T03,
                    ),
                }
                for zone, instance in instances
            ]
            expected_count = (
                spec.offload_min_replicas
                if spec.mode is DeploymentMode.PRODUCTION
                else 1
            )
            expected_trust_mode = (
                "public_system_roots"
                if spec.certificate_strategy is CertificateStrategy.PUBLIC_TRUSTED
                else "presented_chain_pinned"
            )
            t02_passed = len(t02_evidence) >= expected_count and all(
                item.get("status") == 200
                and self._has_hash(item)
                and self._matches_configuration(item, spec)
                for item in t02_evidence
            )
            t03_passed = len(t03_evidence) >= expected_count and all(
                item.get("http_status") == 200
                and item.get("hostname") == spec.private_hostname
                and item.get("trust_mode") == expected_trust_mode
                and item.get("tls_version") in {"TLSv1.2", "TLSv1.3"}
                and isinstance(item.get("subject_alt_names"), list)
                and all(
                    isinstance(name, str)
                    for name in item["subject_alt_names"]
                )
                and spec.private_hostname in item["subject_alt_names"]
                and self._has_hash(item)
                and self._matches_configuration(item, spec)
                for item in t03_evidence
            )
            return [
                self._finding(
                    AcceptanceTestId.T02,
                    t02_passed,
                    "Every offload instance reached the HTTP backend with status 200",
                    "One or more offload-to-backend runtime probes failed",
                    {"instances": t02_evidence},
                ),
                self._finding(
                    AcceptanceTestId.T03,
                    t03_passed,
                    (
                        "Every offload instance passed the configured trust-mode, "
                        "hostname-validating TLS, and HTTP checks"
                    ),
                    "One or more TLS termination runtime probes failed",
                    {"instances": t03_evidence},
                ),
            ]
        except (GoogleApiError, ValueError, KeyError, TypeError):
            return [
                self._failed(
                    AcceptanceTestId.T02,
                    "Offload-to-backend runtime probes could not be verified",
                    "guest-attribute-unavailable",
                ),
                self._failed(
                    AcceptanceTestId.T03,
                    "TLS termination runtime probes could not be verified",
                    "guest-attribute-unavailable",
                ),
            ]

    def _offload_instances(self, spec: DeploymentSpec) -> list[tuple[str, str]]:
        if spec.mode is DeploymentMode.POC:
            return [(spec.zone, f"{spec.name}-offload")]
        url = (
            f"https://compute.googleapis.com/compute/v1/projects/"
            f"{spec.project_id}/regions/{spec.region}/instanceGroupManagers/"
            f"{spec.name}-offload-mig/listManagedInstances"
        )
        managed: list[Any] = []
        seen_tokens: set[str] = set()
        page_token: str | None = None
        complete = False
        page_size = 500
        page_limit = 100
        for _ in range(page_limit):
            body: dict[str, Any] = {"maxResults": page_size}
            if page_token is not None:
                body["pageToken"] = page_token
            _, payload = self._transport.request_json("POST", url, json_body=body)
            page_items = payload.get("managedInstances")
            if not isinstance(page_items, list):
                raise ValueError("invalid-managed-instance-response")
            if len(page_items) > page_size:
                raise ValueError("managed-instance-page-limit")
            managed.extend(page_items)
            if len(managed) > page_size * page_limit:
                raise ValueError("managed-instance-item-limit")
            if "nextPageToken" not in payload or payload["nextPageToken"] == "":
                complete = True
                break
            next_token = payload["nextPageToken"]
            if not isinstance(next_token, str) or next_token in seen_tokens:
                raise ValueError("invalid-managed-instance-page-token")
            seen_tokens.add(next_token)
            page_token = next_token
        if not complete:
            raise ValueError("managed-instance-pagination-incomplete")

        instances: set[tuple[str, str]] = set()
        for item in managed:
            if not isinstance(item, dict) or item.get("instanceStatus") != "RUNNING":
                raise ValueError("managed-instance-not-running")
            instance_url = item.get("instance")
            if not isinstance(instance_url, str):
                raise ValueError("invalid-managed-instance-name")
            parsed = urlsplit(instance_url)
            parts = parsed.path.split("/")
            if (
                parsed.scheme != "https"
                or parsed.hostname not in {"compute.googleapis.com", "www.googleapis.com"}
                or parsed.port is not None
                or parsed.username is not None
                or parsed.password is not None
                or parsed.query
                or parsed.fragment
                or len(parts) != 9
                or parts[1:4] != ["compute", "v1", "projects"]
                or parts[4] != spec.project_id
                or parts[5] != "zones"
                or parts[7] != "instances"
                or not parts[6]
                or not parts[8]
                or "%" in parts[6]
                or "%" in parts[8]
            ):
                raise ValueError("invalid-managed-instance-name")
            identity = (parts[6], parts[8])
            if identity in instances:
                raise ValueError("duplicate-managed-instance")
            instances.add(identity)
        return sorted(instances)

    def _guest_attribute(
        self,
        spec: DeploymentSpec,
        *,
        zone: str,
        instance: str,
        test_id: AcceptanceTestId,
    ) -> dict[str, Any]:
        _, payload = self._transport.request_json(
            "GET",
            (
                f"https://compute.googleapis.com/compute/v1/projects/"
                f"{spec.project_id}/zones/{zone}/instances/{instance}/"
                "getGuestAttributes"
            ),
            params={"queryPath": f"sgstudio/{test_id.value}"},
        )
        value = payload.get("variableValue")
        query_value = payload.get("queryValue")
        if not isinstance(value, str) and isinstance(query_value, dict):
            items = query_value.get("items")
            if isinstance(items, list):
                match = next(
                    (
                        item
                        for item in items
                        if isinstance(item, dict)
                        and item.get("namespace") == "sgstudio"
                        and item.get("key") == test_id.value
                    ),
                    None,
                )
                if isinstance(match, dict):
                    value = match.get("value")
        if not isinstance(value, str):
            raise ValueError("guest-attribute-value-missing")
        decoded = json.loads(value)
        if not isinstance(decoded, dict):
            raise ValueError("guest-attribute-value-invalid")
        return decoded

    def _verify_dns(self, spec: DeploymentSpec) -> AcceptanceFinding:
        try:
            _, address = self._transport.request_json(
                "GET",
                (
                    f"https://compute.googleapis.com/compute/v1/projects/"
                    f"{spec.project_id}/regions/{spec.region}/addresses/"
                    f"{spec.name}-offload-ip"
                ),
            )
            expected = address.get("address")
            if not isinstance(expected, str):
                raise ValueError("internal-address-missing")
            record_name = quote(f"{spec.private_hostname}.", safe="")
            _, record = self._transport.request_json(
                "GET",
                (
                    f"https://dns.googleapis.com/dns/v1/projects/{spec.project_id}/"
                    f"managedZones/{spec.name}-zone/rrsets/{record_name}/A"
                ),
            )
            rrdatas = record.get("rrdatas")
            passed = (
                record.get("name") == f"{spec.private_hostname}."
                and record.get("type") == "A"
                and isinstance(rrdatas, list)
                and rrdatas == [expected]
            )
            return self._finding(
                AcceptanceTestId.T04,
                passed,
                "Private DNS A record exactly matches the reserved internal address",
                "Private DNS does not exactly match the reserved internal address",
                {
                    "hostname": spec.private_hostname,
                    "expected_address": expected,
                    "rrdatas": rrdatas if isinstance(rrdatas, list) else [],
                },
            )
        except (GoogleApiError, ValueError, KeyError, TypeError):
            return self._failed(
                AcceptanceTestId.T04,
                "Private DNS configuration could not be verified",
                "dns-verification-unavailable",
            )

    def _verify_application(self, spec: DeploymentSpec) -> AcceptanceFinding:
        try:
            _, payload = self._transport.request_json(
                "GET",
                (
                    f"https://beyondcorp.googleapis.com/v1/projects/"
                    f"{spec.project_id}/locations/global/securityGateways/"
                    f"{spec.gateway_id}/applications/{spec.name}-app"
                ),
            )
            matchers = payload.get("endpointMatchers", payload.get("endpoint_matchers"))
            exact = (
                isinstance(matchers, list)
                and len(matchers) == 1
                and isinstance(matchers[0], dict)
                and matchers[0].get("hostname") == spec.application_hostname
                and matchers[0].get("ports") == [spec.application_port]
            )
            upstreams = payload.get("upstreams")
            network_name = (
                spec.vpc_name
                if spec.network_strategy.value == "existing"
                else f"{spec.name}-vpc"
            )
            expected_network = (
                f"projects/{spec.upstream_project_id}/global/networks/{network_name}"
            )
            upstream_exact = False
            if isinstance(upstreams, list) and len(upstreams) == 1:
                upstream = upstreams[0]
                if isinstance(upstream, dict):
                    network = upstream.get("network")
                    policy = upstream.get("egressPolicy", upstream.get("egress_policy"))
                    expected_policy = (
                        {"regions": [spec.application_egress_region]}
                        if spec.application_egress_region
                        else None
                    )
                    upstream_exact = (
                        isinstance(network, dict)
                        and network.get("name") == expected_network
                        and policy == expected_policy
                    )
            exact = exact and upstream_exact
            return self._finding(
                AcceptanceTestId.T05,
                exact,
                "Secure Gateway application exactly matches the private HTTPS endpoint",
                "Secure Gateway application matcher is missing or does not match exactly",
                {
                    "application": f"{spec.name}-app",
                    "hostname": spec.application_hostname,
                    "port": spec.application_port,
                    "network": expected_network,
                    "egress_region": spec.application_egress_region,
                    "exact_match": exact,
                },
            )
        except (GoogleApiError, ValueError, KeyError, TypeError):
            return self._failed(
                AcceptanceTestId.T05,
                "Secure Gateway application matcher could not be verified",
                "application-verification-unavailable",
            )

    @staticmethod
    def _has_hash(evidence: dict[str, Any]) -> bool:
        value = evidence.get("body_sha256")
        return (
            isinstance(value, str)
            and len(value) == 64
            and all(character in "0123456789abcdef" for character in value)
        )

    @staticmethod
    def _matches_configuration(
        evidence: dict[str, Any],
        spec: DeploymentSpec,
    ) -> bool:
        expected = canonical_configuration_hash(spec)
        return evidence.get("configuration_hash") == expected

    @classmethod
    def _finding(
        cls,
        test_id: AcceptanceTestId,
        passed: bool,
        passed_summary: str,
        failed_summary: str,
        evidence: dict[str, Any],
    ) -> AcceptanceFinding:
        return AcceptanceFinding(
            test_id=test_id,
            status=(AcceptanceStatus.PASSED if passed else AcceptanceStatus.FAILED),
            summary=passed_summary if passed else failed_summary,
            evidence=canonical_json(evidence),
        )

    @classmethod
    def _failed(
        cls,
        test_id: AcceptanceTestId,
        summary: str,
        reason: str,
    ) -> AcceptanceFinding:
        return cls._finding(
            test_id,
            False,
            summary,
            summary,
            {"reason": reason},
        )


def create_google_acceptance_verifier(
    *, require_impersonation: bool = False
) -> GoogleAcceptanceVerifier:
    return GoogleAcceptanceVerifier(
        GoogleAuthorizedTransport.from_adc(require_impersonation=require_impersonation)
    )

File: secure-gateway-studio/extension/src/providers/acceptance.ts
/**
 * Acceptance verification shared with the canonical T01-T09 matrix.
 *
 * Machine checks are deliberately fail-closed. Browser observations remain
 * operator evidence; the extension never turns a successful API fixture into
 * a claim that a managed browser actually reached the application.
 */

import { configurationHash } from "../domain/planner.ts";
import {
  applicationHostname,
  applicationPort,
  upstreamProjectId,
  type ChromePlatform,
  type DeploymentSpec,
} from "../domain/spec.ts";
import type { Transport } from "./executor.ts";

export type AcceptanceStatus =
  | "passed"
  | "failed"
  | "pending"
  | "skipped"
  | "user_confirmed";
export type EvidenceSource = "system_verified" | "operator_confirmed";

export interface AcceptanceRequirement {
  test_id: string;
  case_key: string;
  title: string;
  source: EvidenceSource;
  operator_confirmable: boolean;
  applicable: true;
  /** Only the T06 regression control may be skipped by a greenfield PoC. */
  allow_poc_skip: boolean;
  detail: string;
}

export interface AcceptanceFinding {
  test_id: string;
  status: AcceptanceStatus;
  summary: string;
  evidence: string;
}

const TITLES: Record<string, string> = {
  T01: "HTTP backend local response",
  T02: "Offload-to-backend response",
  T03: "TLS termination",
  T04: "Private DNS",
  T05: "Secure Gateway route",
  T06: "Existing HTTPS regression control",
  T07: "Managed Chrome end to end",
  T08: "Log correlation",
  T09: "Unauthorized or unmanaged denial",
};

const MANAGED_INSTANCE_PAGE_SIZE = 500;
const MAX_MANAGED_INSTANCE_PAGES = 100;
const MAX_MANAGED_INSTANCE_ITEMS =
  MANAGED_INSTANCE_PAGE_SIZE * MAX_MANAGED_INSTANCE_PAGES;
const COMPUTE_SELF_LINK_HOSTS = new Set([
  "compute.googleapis.com",
  "www.googleapis.com",
]);

function requirement(
  testId: string,
  source: EvidenceSource,
  caseKey = "default",
  detail = "",
  allowPocSkip = false,
): AcceptanceRequirement {
  return {
    test_id: testId,
    case_key: caseKey,
    title: TITLES[testId],
    source,
    operator_confirmable: source === "operator_confirmed",
    applicable: true,
    allow_poc_skip: allowPocSkip,
    detail,
  };
}

/** Return the exact required cases defined by docs/TEST_MATRIX.md. */
export function acceptanceRequirements(spec: DeploymentSpec): AcceptanceRequirement[] {
  const system = "system_verified" as const;
  const operator = "operator_confirmed" as const;
  const requirements: AcceptanceRequirement[] = [];

  if (spec.backend_kind === "direct_https") {
    requirements.push(requirement("T05", system));
  } else if (spec.backend_kind === "internal_https_lb") {
    requirements.push(
      requirement("T01", system),
      requirement("T03", operator, "default", "Verify TLS at the ILB endpoint."),
      requirement("T04", system),
      requirement("T05", system),
    );
  } else {
    requirements.push(
      requirement(
        "T01",
        spec.backend_kind === "existing_http" ? operator : system,
        "default",
        spec.backend_kind === "existing_http"
          ? "Record the existing backend's known-good local response."
          : "Read the managed backend runtime probe.",
      ),
      requirement("T02", system),
      requirement("T03", system),
      requirement("T04", system),
      requirement("T05", system),
    );
  }

  requirements.push(
    requirement(
      "T06",
      operator,
      "default",
      "Verify an already-known-good HTTPS control application; a greenfield PoC may skip with a reason.",
      true,
    ),
  );

  const platformOrder: ChromePlatform[] = ["macos", "windows", "linux", "chromeos"];
  for (const platform of platformOrder) {
    if (spec.platforms.has(platform)) {
      requirements.push(
        requirement("T07", operator, platform, `Verify managed Chrome on ${platform}.`),
      );
    }
  }

  if (spec.mode === "production") {
    requirements.push(
      requirement("T08", operator),
      requirement("T09", operator, "unauthorized-principal"),
      requirement("T09", operator, "unmanaged-browser"),
    );
  }
  return requirements;
}

const BEYONDCORP = "https://beyondcorp.googleapis.com/v1";
const COMPUTE = "https://compute.googleapis.com/compute/v1";
const DNS = "https://dns.googleapis.com/dns/v1";

function jsonEvidence(runId: string, value: Record<string, unknown>): string {
  return JSON.stringify({ run_id: runId, ...value });
}

function hasSha256(value: unknown): boolean {
  return typeof value === "string" && /^[0-9a-f]{64}$/.test(value);
}

function passedConfiguration(evidence: Record<string, unknown>, spec: DeploymentSpec): boolean {
  return evidence.configuration_hash === configurationHash(spec);
}

export class GoogleAcceptanceVerifier {
  private readonly transport: Transport;

  constructor(transport: Transport) {
    this.transport = transport;
  }

  /** Run every applicable system-verifiable check, never T06-T09. */
  async verify(spec: DeploymentSpec, runId: string): Promise<AcceptanceFinding[]> {
    const findings: AcceptanceFinding[] = [];
    if (spec.backend_kind === "managed_sample" || spec.backend_kind === "internal_https_lb") {
      findings.push(await this.verifyBackend(spec, runId));
    }
    if (spec.backend_kind !== "direct_https" && spec.backend_kind !== "internal_https_lb") {
      findings.push(...(await this.verifyOffload(spec, runId)));
    }
    if (spec.backend_kind !== "direct_https") {
      findings.push(await this.verifyDns(spec, runId));
    }
    findings.push(await this.verifyApplication(spec, runId));
    return findings;
  }

  private async guestAttribute(
    spec: DeploymentSpec,
    zone: string,
    instance: string,
    testId: string,
  ): Promise<Record<string, unknown>> {
    const { payload } = await this.transport.requestJson(
      "GET",
      `${COMPUTE}/projects/${spec.project_id}/zones/${zone}/instances/${instance}/getGuestAttributes`,
      { params: { queryPath: `sgstudio/${testId}` } },
    );
    let value = payload.variableValue;
    const queryValue = payload.queryValue as Record<string, unknown> | undefined;
    if (typeof value !== "string" && queryValue && Array.isArray(queryValue.items)) {
      const item = queryValue.items.find((entry) => {
        const record = entry as Record<string, unknown>;
        return record.namespace === "sgstudio" && record.key === testId;
      }) as Record<string, unknown> | undefined;
      value = item?.value;
    }
    if (typeof value !== "string") throw new Error("guest-attribute-value-missing");
    const decoded = JSON.parse(value) as unknown;
    if (decoded === null || typeof decoded !== "object" || Array.isArray(decoded)) {
      throw new Error("guest-attribute-value-invalid");
    }
    return decoded as Record<string, unknown>;
  }

  private async verifyBackend(spec: DeploymentSpec, runId: string): Promise<AcceptanceFinding> {
    try {
      const evidence = await this.guestAttribute(
        spec,
        spec.zone,
        `${spec.name}-backend`,
        "T01",
      );
      const passed =
        evidence.status === 200 &&
        hasSha256(evidence.body_sha256) &&
        passedConfiguration(evidence, spec);
      return {
        test_id: "T01",
        status: passed ? "passed" : "failed",
        summary: passed
          ? "Managed backend returned HTTP 200 to its local runtime probe."
          : "Managed backend runtime probe is missing or invalid.",
        evidence: jsonEvidence(runId, evidence),
      };
    } catch (error) {
      return this.failed("T01", runId, "Managed backend runtime probe could not be verified.", error);
    }
  }

  private async offloadInstances(spec: DeploymentSpec): Promise<Array<[string, string]>> {
    if (spec.mode === "poc") return [[spec.zone, `${spec.name}-offload`]];
    const url =
      `${COMPUTE}/projects/${spec.project_id}/regions/${spec.region}/instanceGroupManagers/` +
      `${spec.name}-offload-mig/listManagedInstances`;
    const managed: unknown[] = [];
    const seenTokens = new Set<string>();
    let pageToken: string | undefined;
    let complete = false;
    for (let page = 0; page < MAX_MANAGED_INSTANCE_PAGES; page += 1) {
      const jsonBody: Record<string, unknown> = { maxResults: MANAGED_INSTANCE_PAGE_SIZE };
      if (pageToken !== undefined) jsonBody.pageToken = pageToken;
      const { payload } = await this.transport.requestJson("POST", url, { jsonBody });
      if (!Array.isArray(payload.managedInstances)) {
        throw new Error("invalid-managed-instance-response");
      }
      if (payload.managedInstances.length > MANAGED_INSTANCE_PAGE_SIZE) {
        throw new Error("managed-instance-page-limit");
      }
      managed.push(...payload.managedInstances);
      if (managed.length > MAX_MANAGED_INSTANCE_ITEMS) {
        throw new Error("managed-instance-item-limit");
      }
      if (!("nextPageToken" in payload) || payload.nextPageToken === "") {
        complete = true;
        break;
      }
      if (
        typeof payload.nextPageToken !== "string" ||
        seenTokens.has(payload.nextPageToken)
      ) {
        throw new Error("invalid-managed-instance-page-token");
      }
      seenTokens.add(payload.nextPageToken);
      pageToken = payload.nextPageToken;
    }
    if (!complete) throw new Error("managed-instance-pagination-incomplete");

    const instances = new Map<string, [string, string]>();
    for (const entry of managed) {
      if (entry === null || typeof entry !== "object" || Array.isArray(entry)) {
        throw new Error("invalid-managed-instance-response");
      }
      const record = entry as Record<string, unknown>;
      if (record.instanceStatus !== "RUNNING" || typeof record.instance !== "string") {
        throw new Error("managed-instance-not-running");
      }
      let parsed: URL;
      try {
        parsed = new URL(record.instance);
      } catch {
        throw new Error("invalid-managed-instance-name");
      }
      const parts = parsed.pathname.split("/");
      if (
        parsed.protocol !== "https:" ||
        !COMPUTE_SELF_LINK_HOSTS.has(parsed.hostname) ||
        parsed.port !== "" || parsed.username !== "" || parsed.password !== "" ||
        parsed.search !== "" || parsed.hash !== "" ||
        parts.length !== 9 || parts[1] !== "compute" || parts[2] !== "v1" ||
        parts[3] !== "projects" || parts[4] !== spec.project_id ||
        parts[5] !== "zones" || parts[7] !== "instances" ||
        !parts[6] || !parts[8] ||
        decodeURIComponent(parts[6]) !== parts[6] || decodeURIComponent(parts[8]) !== parts[8]
      ) {
        throw new Error("invalid-managed-instance-name");
      }
      const zone = parts[6];
      const instance = parts[8];
      const key = `${zone}/${instance}`;
      if (instances.has(key)) throw new Error("duplicate-managed-instance");
      instances.set(key, [zone, instance]);
    }
    return [...instances.values()].sort((left, right) => left.join("/").localeCompare(right.join("/")));
  }

  private async verifyOffload(
    spec: DeploymentSpec,
    runId: string,
  ): Promise<AcceptanceFinding[]> {
    try {
      const instances = await this.offloadInstances(spec);
      if (instances.length === 0) throw new Error("no-offload-instances");
      const t02: Record<string, unknown>[] = await Promise.all(
        instances.map(async ([zone, instance]) => ({
          instance,
          ...(await this.guestAttribute(spec, zone, instance, "T02")),
        })),
      );
      const t03: Record<string, unknown>[] = await Promise.all(
        instances.map(async ([zone, instance]) => ({
          instance,
          ...(await this.guestAttribute(spec, zone, instance, "T03")),
        })),
      );
      const expectedCount = spec.mode === "production" ? spec.offload_min_replicas : 1;
      const expectedTrustMode = spec.certificate_strategy === "public_trusted"
        ? "public_system_roots"
        : "presented_chain_pinned";
      const t02Passed =
        t02.length >= expectedCount &&
        t02.every(
          (item) =>
            item.status === 200 &&
            hasSha256(item.body_sha256) &&
            passedConfiguration(item, spec),
        );
      const t03Passed =
        t03.length >= expectedCount &&
        t03.every(
          (item) =>
            item.http_status === 200 &&
            item.hostname === spec.private_hostname &&
            item.trust_mode === expectedTrustMode &&
            (item.tls_version === "TLSv1.2" || item.tls_version === "TLSv1.3") &&
            Array.isArray(item.subject_alt_names) &&
            item.subject_alt_names.every((name) => typeof name === "string") &&
            item.subject_alt_names.includes(spec.private_hostname) &&
            hasSha256(item.body_sha256) &&
            passedConfiguration(item, spec),
        );
      return [
        {
          test_id: "T02",
          status: t02Passed ? "passed" : "failed",
          summary: t02Passed
            ? "Every offload instance reached the HTTP backend."
            : "One or more offload-to-backend runtime probes failed.",
          evidence: jsonEvidence(runId, { instances: t02 }),
        },
        {
          test_id: "T03",
          status: t03Passed ? "passed" : "failed",
          summary: t03Passed
            ? "Every offload instance passed the configured trust-mode, hostname-validating TLS, and HTTP checks."
            : "One or more TLS termination runtime probes failed.",
          evidence: jsonEvidence(runId, { instances: t03 }),
        },
      ];
    } catch (error) {
      return [
        this.failed("T02", runId, "Offload-to-backend probes could not be verified.", error),
        this.failed("T03", runId, "TLS termination probes could not be verified.", error),
      ];
    }
  }

  private async verifyDns(spec: DeploymentSpec, runId: string): Promise<AcceptanceFinding> {
    try {
      const { payload: address } = await this.transport.requestJson(
        "GET",
        `${COMPUTE}/projects/${spec.project_id}/regions/${spec.region}/addresses/${spec.name}-offload-ip`,
      );
      if (typeof address.address !== "string") throw new Error("internal-address-missing");
      const recordName = encodeURIComponent(`${spec.private_hostname}.`);
      const { payload: record } = await this.transport.requestJson(
        "GET",
        `${DNS}/projects/${spec.project_id}/managedZones/${spec.name}-zone/rrsets/${recordName}/A`,
      );
      const passed =
        record.name === `${spec.private_hostname}.` &&
        record.type === "A" &&
        Array.isArray(record.rrdatas) &&
        record.rrdatas.length === 1 &&
        record.rrdatas[0] === address.address;
      return {
        test_id: "T04",
        status: passed ? "passed" : "failed",
        summary: passed
          ? "Private DNS exactly matches the reserved internal address."
          : "Private DNS does not exactly match the reserved internal address.",
        evidence: jsonEvidence(runId, {
          hostname: spec.private_hostname,
          expected_address: address.address,
          rrdatas: Array.isArray(record.rrdatas) ? record.rrdatas : [],
        }),
      };
    } catch (error) {
      return this.failed("T04", runId, "Private DNS could not be verified.", error);
    }
  }

  private async verifyApplication(
    spec: DeploymentSpec,
    runId: string,
  ): Promise<AcceptanceFinding> {
    try {
      const applicationUrl =
        `${BEYONDCORP}/projects/${spec.project_id}/locations/global` +
        `/securityGateways/${spec.gateway_id}/applications/${spec.name}-app`;
      const { payload } = await this.transport.requestJson("GET", applicationUrl);
      const matchers = payload.endpointMatchers ?? payload.endpoint_matchers;
      const expectedHost = applicationHostname(spec);
      const expectedPort = applicationPort(spec);
      const matcherExact =
        Array.isArray(matchers) &&
        matchers.length === 1 &&
        (() => {
          const record = matchers[0] as Record<string, unknown>;
          return record.hostname === expectedHost &&
            Array.isArray(record.ports) &&
            record.ports.length === 1 &&
            Number(record.ports[0]) === expectedPort;
        })();
      const networkName =
        spec.network_strategy === "existing" ? spec.vpc_name : `${spec.name}-vpc`;
      const expectedNetwork =
        `projects/${upstreamProjectId(spec)}/global/networks/${networkName}`;
      const upstreams = Array.isArray(payload.upstreams) ? payload.upstreams : [];
      const upstreamExact = upstreams.length === 1 && (() => {
        const upstream = upstreams[0] as Record<string, unknown>;
        const network = upstream.network as Record<string, unknown> | undefined;
        if (network?.name !== expectedNetwork) return false;
        const policy = (upstream.egressPolicy ?? upstream.egress_policy) as
          | Record<string, unknown>
          | undefined;
        if (!spec.application_egress_region) return policy === undefined;
        return (
          Array.isArray(policy?.regions) &&
          policy.regions.length === 1 &&
          policy.regions[0] === spec.application_egress_region
        );
      })();
      const passed = matcherExact && upstreamExact;
      return {
        test_id: "T05",
        status: passed ? "passed" : "failed",
        summary: passed
          ? "Secure Gateway application exactly matches the intended route."
          : "Secure Gateway matcher or upstream network does not match exactly.",
        evidence: jsonEvidence(runId, {
          application: `${spec.name}-app`,
          hostname: expectedHost,
          port: expectedPort,
          network: expectedNetwork,
          egress_region: spec.application_egress_region,
          exact_match: passed,
        }),
      };
    } catch (error) {
      return this.failed("T05", runId, "Secure Gateway application could not be verified.", error);
    }
  }

  private failed(
    testId: string,
    runId: string,
    summary: string,
    error: unknown,
  ): AcceptanceFinding {
    return {
      test_id: testId,
      status: "failed",
      summary,
      evidence: jsonEvidence(runId, {
        error: error instanceof Error ? error.message : String(error),
      }),
    };
  }
}
